### PR TITLE
[5.1] Create access responses and unauthorized exception

### PR DESCRIPTION
**Pull request by JosephSilber**
This is a different take on #10132.

---
### Usage

``` php
use Illuminate/Auth/Access/HandlesAuthorization;

class PostPolicy
{
    use HandlesAuthorization;

    public function update(User $user, Post $post)
    {
        if ($user->isSubscriber()) {
            return $this->deny('Subscribers cannot edit posts');
        }

        if ($user->id != $post->user_id) {
            return $this->deny('You do not own this post');
        }

        return $this->allow($user->type().'s can edit their own posts');
    }
}
```

The `allow` call might be a contrived example, but it's even more useful for example from within a `before` filter and such.

``` php
// @throws \Illuminate\Auth\Access\HandlesAuthorization
$response = Gate::authorize('update', $post);

Log::info('Admission granted: '.$response->message());
```

You can just ignore the return value if you want, just like using `AuthorizesRequests` in the controller:

``` php
// @throws \Illuminate\Auth\Access\HandlesAuthorization
Gate::authorize('update', $post);

// user is authorized
```

### CODE REVIEW

**Comment by garygreen** [comment]
Not a fan of this implementation at all. Much prefer the other with its single return object. There's so many weird paths this is taking whereas the other one felt much simpler and easier to understand.

**Comment by lucasmichot** [comment]
@JosephSilber, it would make sense IMO to make `\Illuminate\Auth\Access\UnautorizedException` extends `\Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException` and then construct it without the `403` HTTP error code.

File: src/Illuminate/Auth/Access/UnautorizedException.php [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Illuminate\Auth\Access;
+
+use Exception;
+
+class UnautorizedException extends Exception
```
Comment (lucasmichot) [inline src/Illuminate/Auth/Access/UnautorizedException.php]: typo

**Comment by JosephSilber** [comment]
@lucasmichot I intentionally avoided that. Unlike the `AuthorizesRequests` trait, gate authorization is not limited to controllers or HTTP requests. You can call `Gate::authorize()` from within a queue job or even an artisan command. The gate should not be aware of the HTTP layer at all.

**Comment by RobvH** [comment]
I'm growing to like "HandlesAuthorization"

**Comment by garygreen** [comment]
Yippee!! <3

**Comment by ninjaparade** [comment]
:boom: :boom::boom::boom::boom: :gun: :dancer: :dancer: :dancer: :dancer: :dancer: :dancer:

**Pull request merged**